### PR TITLE
Move terminal to custom hooks

### PR DESCRIPTION
**Pull request by rikasun**
known issues: 
https://linear.app/cased/issue/CAS-475/cant-init-another-terminal-unless-page-reload
https://linear.app/cased/issue/CAS-476/terminal-fitaddonpossibly-not-working-properly-on-launch

### CODE REVIEW

**Comment by rikasun** [comment]
@ashblue I did a bunch of hack to get the test pass because of https://linear.app/cased/issue/CAS-475/cant-init-another-terminal-unless-page-reload
Let me know if you have any better solutions! Thanks!

File: libs/redux/src/lib/stores/prompt/prompt.store.ts [hunk 1]
```diff
@@ -165,6 +169,10 @@ export const promptStore: IPromptStore = {
         return;
       }
 
+      const term = getStoreState().terminal.terminal;
+      // istanbul ignore next
+      if (!term) return;
```
Comment (ashblue) [inline libs/redux/src/lib/stores/prompt/prompt.store.ts:174]: These lines here seem odd. Does a test need a default value in the store for terminal maybe?

File: apps/shell/src/app/modules/prompt/prompt.tsx [hunk 2]
```diff
@@ -17,7 +9,14 @@ import PromptSearch from './search/prompt-search';
...
 
-let terminalMounted = false;
+// hack for testing
+declare global {
+  interface Window {
+    terminalMounted: boolean;
+  }
+}
+
+window.terminalMounted = false;
```
Comment (ashblue) [inline apps/shell/src/app/modules/prompt/prompt.tsx:19]: Should be a work around to remove this in the test suite. Probably just need to adjust the factory class generator.

File: apps/shell/src/app/modules/prompt/prompt.spec.tsx [hunk 3]
```diff
@@ -110,6 +111,14 @@ const setup = (options: IOptions = {}) => {
 };
 
 describe('Prompt', () => {
+  // @NOTE This is a hack to allow terminal mount each session
+  beforeEach(() => {
+    Object.defineProperty(window, 'terminalMounted', {
+      value: false,
+      writable: true,
+    });
+  });
```
Comment (ashblue) [inline apps/shell/src/app/modules/prompt/prompt.spec.tsx:120]: I think we can just spy on the factory for the terminal classes and avoid this.

**Review by ashblue — CHANGES_REQUESTED** [review]
This looks great. The component is much cleaner and easier to read. I think we can remove a lot of the extra code I flagged with some spies on factory generator. Also I noticed some memory leaks caused by scoping on the window resize event that should be easy to clean up.

PS Happy to do a pair to tackle the tagged extra testing blocks. Removing them will get rid of the remaining hacks.

**Review by ashblue — APPROVED** [review]
Rika and I paired for a significant refactor on the code to add a few abstraction layers. It removes a large number of if statements, centralizes xterm usage, and allows for custom terminal line writing commands.

**Pull request merged**